### PR TITLE
:seedling: Test flow tweaks

### PR DESCRIPTION
**Pull request by dhaiducek**
- chore: use `go.mod` for workflow Go version
- chore: use local bin rather than `/tmp` for CLIs
- chore: use `setup-envtest` CLI for envtest
- chore: bump Kind cluster K8s version
- chore: simplify E2E command
- chore: pass streams to Helm client
- chore: update flows for flaky spoke cluster deletion

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved join and addon lifecycle cleanup to reduce test failures and ensure resources are removed in the correct order.
  - Improved handling of cluster deletion, including stalled resources.
  - Corrected error handling when hub resource discovery fails.
  - Ensured command output is routed consistently during Helm operations.

- **Tests**
  - Updated end-to-end and integration test setup, fixtures, paths, and readiness checks.
  - Go and Kubernetes tool versions now align automatically with project configuration.
  - Improved local test tool setup and environment configuration.

- **Chores**
  - Strengthened CI workflow permissions and action security.
  - Pinned CI actions and end-to-end test infrastructure for more consistent runs.

### CODE REVIEW

File: pkg/helpers/helm/helm.go [hunk 1]
```diff
@@ -146,7 +149,7 @@ func (h *Helm) PrepareChart(repoName, repoURL string) error {
 		if err := f.WriteFile(repoFile, 0644); err != nil {
 			log.Fatal(err)
 		}
-		fmt.Printf("%q has been added to your repositories\n", repoName)
+		fmt.Fprintf(h.streams.Out, "%q has been added to your repositories\n", repoName)
```
[the remark this answers is not in the document]
Comment (dhaiducek) [reply pkg/helpers/helm/helm.go:152]: This can be tech debt since it's the pattern that exists currently throughout the repo.

File: test/e2e/e2e-test.mk [hunk 2]
```diff
@@ -17,15 +17,12 @@ clean-e2e:
...
-	kind create cluster --name ${HUB_NAME} --image kindest/node:v1.24.0
-.PHONY: start-cluster 
+	kind create cluster --name ${HUB_NAME} --image kindest/node:v1.35.5
+.PHONY: start-cluster
 
-test-e2e: clean-e2e envtest-setup ensure-ginkgo start-cluster deps install
-	go test -c ./test/e2e/clusteradm
-	./clusteradm.test -test.v -ginkgo.v -test.timeout=3600s \
-	$(if $(GINKGO_LABEL_FILTER),-ginkgo.label-filter="$(GINKGO_LABEL_FILTER)") -bundle-version=$(BUNDLE_VERSION)
+test-e2e: clean-e2e ensure-ginkgo start-cluster deps install test-e2e-only
```
[the remark this answers is not in the document]
Comment (dhaiducek) [reply test/e2e/e2e-test.mk:23]: This is unlikely since we don't use the `-j` flag in our CI.

File: pkg/helpers/helm/helm.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dhaiducek) [reply pkg/helpers/helm/helm.go:152]: No issue for now, thanks. It's not currently high priority.

File: test/e2e/clusteradm/joinhubscenario_grpc_test.go [hunk 3]
```diff
@@ -21,13 +21,13 @@ var _ = ginkgo.Describe("test clusteradm join with grpc", ginkgo.Label("join-hub
 		gomega.Expect(err).NotTo(gomega.HaveOccurred())
 	})
 	ginkgo.AfterEach(func() {
-		err := util.WaitClustersDeleted(e2e.Cluster().Hub().KubeConfig())
+		err := e2e.Clusteradm().Unjoin(
+			"--context", e2e.Cluster().ManagedCluster1().Context(),
+			"--cluster-name", e2e.Cluster().ManagedCluster1().Name(),
+		)
```
[the remark this answers is not in the document]
Comment (dhaiducek) [reply test/e2e/clusteradm/joinhubscenario_grpc_test.go:27]: Check this again.

**Review by mikeshng — COMMENTED** [review]
Need rebase and @TylerGillson will have to approve in case it breaks fleet-config.

File: test/e2e/util/helper.go [hunk 4]
```diff
@@ -63,24 +64,39 @@ func WaitClustersDeleted(restcfg *rest.Config) error {
...
+			}
+
+			// The registration controller can get stuck removing a cluster from the
+			// default ManagedClusterSet when DefaultClusterSet is enabled. Unblock
+			// e2e cleanup after giving the controller time to finish.
+			if len(mcl.Finalizers) > 0 && time.Since(mcl.DeletionTimestamp.Time) > 60*time.Second {
+				mcl.Finalizers = nil
+				if _, err := clientset.ClusterV1().ManagedClusters().Update(ctx, &mcl, metav1.UpdateOptions{}); err != nil && !errors.IsNotFound(err) {
+					return false, err
+				}
```
Comment (TylerGillson) [inline test/e2e/util/helper.go]: This is a band-aid solution. Is there an issue filed to address this properly in the registration controller?

File: pkg/cmd/install/hubaddon/exec_test.go [hunk 5]
```diff
@@ -112,14 +113,15 @@ var _ = ginkgo.Describe("install hub-addon", func() {
 		addon := "argocd"
 		clusteradmFlagsCopy := *clusteradmFlags
 		clusteradmFlagsCopy.DryRun = true
+		streams := genericiooptions.IOStreams{Out: io.Discard, ErrOut: os.Stderr}
```
Comment (TylerGillson) [inline pkg/cmd/install/hubaddon/exec_test.go:116]: This is a functional change (`Out: io.Discard` instead of `os.Stdout`).

@mikeshng we really want to suppress all helm output?

Comment (mikeshng) [reply pkg/cmd/install/hubaddon/exec_test.go:116]: @dhaiducek This is a test file so it should be ok to not change it? PTAL Thanks!

Comment (dhaiducek) [reply pkg/cmd/install/hubaddon/exec_test.go:116]: For the test here, printing the output during dryrun is very noisy since it's all of the manifests in the chart and is huge.

File: test/e2e/util/helper.go (hunk 4 above)
Comment (dhaiducek) [reply test/e2e/util/helper.go]: Hmmm. Good point. I'll open an issue in `ocm`. Maybe we should we just leave this off, then?

File: pkg/cmd/install/hubaddon/exec_test.go (hunk 5 above)
Comment (TylerGillson) [reply pkg/cmd/install/hubaddon/exec_test.go:116]: Sorry I reviewed too quickly... thought it was not test code.

File: test/e2e/util/helper.go (hunk 4 above)
Comment (TylerGillson) [reply test/e2e/util/helper.go]: @mikeshng wdyt? Personally I'd be fine with adding a link to the issue + a TODO to remove it once resolved.

Comment (mikeshng) [reply test/e2e/util/helper.go]: SGTM! Thank you both!

Comment (dhaiducek) [reply test/e2e/util/helper.go]: I engaged the Cursor agent about this. This code path isn't executed frequently in the test (only twice as far as I can tell). If the DefaultClusterSet feature gate were enabled _after_ a managed cluster was attached it'd be missing the clusterset label and could trigger this scenario. Or else maybe the webhook wasn't running yet and didn't label the managed cluster properly. In either case, considering the DefaultClusterSet feature gate is usually enabled by default from the get-go, it seems this might be an odd edge case or infrastructure issue. I'm just going to remove this code for now and we can address it if we see it again in the future.

Comment (dhaiducek) [reply test/e2e/util/helper.go]: Updated--thanks for calling it out, @TylerGillson!

**Review by TylerGillson — APPROVED** [review]

File: test/e2e/util/util.go [hunk 6]
```diff
@@ -89,14 +89,13 @@ func initE2E(version string) (*TestE2eConfig, error) {
 		fmt.Println("cleaning hub...")
 
 		fmt.Println("deleting all clusters on the hub...")
-		// delete all clusters
-		err = WaitClustersDeleted(e2eConf.Cluster().Hub().KubeConfig())
+		err := WaitClustersDeleted(e2eConf.Cluster().Hub().KubeConfig())
```
[the remark this answers is not in the document]
Comment (dhaiducek) [reply test/e2e/util/util.go:92]: Why is that the "required" cleanup order?

File: test/e2e/util/helper.go (hunk 4 above)
Comment (dhaiducek) [reply test/e2e/util/helper.go]: The band-aid comment from @TylerGillson was an honest truth I couldn't let go, and I didn't really want a workaround here if the fix might be feasible beforehand. I think I've narrowed down the cause of the finalizer removal failure as the `DefaultClusterSet` feature gate. I've opened a PR here and will retry the CI in this PR once it merges to see whether we're on a more stable path:

- https://github.com/open-cluster-management-io/ocm/pull/1692

**Comment by dhaiducek** [comment]
After merging the `ocm` PR, I re-ran the CI here, and it passed.
So I re-re-ran it...and it passed again. 🤷‍♂️ 
I'm tempted to run it again since I don't quite believe it, but I think it's safe to say this PR is ready for review. 🥳 😂 🎉 

@TylerGillson @mikeshng

**Review by mikeshng — APPROVED** [review]
/lgtm

Thank you!
